### PR TITLE
Introduction of Fuid property type

### PR DESCRIPTION
**Pull request by adamfox-nanopay**
Adds a new Fuid property type

### CODE REVIEW

File: src/foam/core/AbstractFuidPropertyInfo.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/core/AbstractFuidPropertyInfo.java]: 2021

File: src/foam/core/types.js [hunk 2]
```diff
@@ -1091,3 +1091,30 @@ foam.CLASS({
...
+foam.CLASS({
+  package: 'foam.core',
+  name: 'Fuid',
+  extends: 'Property',
+
+  properties: [
+    {
+      name: 'adapt',
+      value: function(_, a, p) {
+        if ( foam.Object.isInstance(a) ) {
```
Comment (kgrgreer) [inline src/foam/core/types.js]: I don't think we want to translate FUIDs so this part can be removed.

File: src/foam/core/types.js [hunk 3]
```diff
@@ -1091,3 +1091,30 @@ foam.CLASS({
...
+          if ( a[foam.locale] !== undefined )
+            return a[foam.locale];
+          if ( a[foam.locale.substring(0, foam.locale.indexOf('-'))] !== undefined )
+            return a[foam.locale.substring(0, foam.locale.indexOf('-'))];
+          return a['en'];// default language.
+        }
+        var s = typeof a === 'function' ||
+                typeof a === 'number'   ? String(a)                :
```
Comment (kgrgreer) [inline src/foam/core/types.js]: Just check if a number and convert to a String, then trim.

File: src/foam/core/AbstractFuidPropertyInfo.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+package foam.core;
+
+import javax.xml.stream.XMLStreamReader;
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.security.Signature;
+import java.security.SignatureException;
+
+public abstract class AbstractFuidPropertyInfo
+    extends AbstractPropertyInfo
```
Comment (jlhughes) [inline src/foam/core/AbstractFuidPropertyInfo.java]: Why not extent AbstractLongPropertyInfo?

File: src/foam/core/types.js [hunk 5]
```diff
@@ -1091,3 +1091,30 @@ foam.CLASS({
     }
   ]
 });
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'Fuid',
```
Comment (kgrgreer) [inline src/foam/core/types.js]: How about 'FUIDProperty'?

File: src/foam/java/refinements.js [hunk 6]
```diff
@@ -2500,3 +2500,16 @@ foam.CLASS({
...
+
+foam.CLASS({
+  package: 'foam.java',
+  name: 'FuidJavaRefinement',
+  refines: 'foam.core.Fuid',
+  flags: ['java'],
+  mixins: [ 'foam.java.JavaCompareImplementor' ],
+
+  properties: [
+    ['javaType',     'Long'],
```
Comment (lchanmann) [inline src/foam/java/refinements.js]:
```suggestion
    ['javaType',     'long'],
```

File: src/foam/java/refinements.js [hunk 7]
```diff
@@ -2500,3 +2500,17 @@ foam.CLASS({
...
+  package: 'foam.java',
+  name: 'FUIDJavaRefinement',
+  refines: 'foam.core.FUIDProperty',
+  flags: ['java'],
+  mixins: [ 'foam.java.JavaCompareImplementor' ],
+
+  properties: [
+    ['javaType',     'long'],
+    ['value',        0],
+    ['javaInfoType', 'foam.core.AbstractFUIDPropertyInfo']
```
Comment (lchanmann) [inline src/foam/java/refinements.js:2514]:
```suggestion
    ['javaInfoType', 'foam.core.AbstractLongPropertyInfo']
```

Comment (lchanmann) [reply src/foam/java/refinements.js:2514]: please disregard

**Review by lchanmann — APPROVED** [review]

File: src/foam/core/types.js [hunk 8]
```diff
@@ -1091,3 +1091,23 @@ foam.CLASS({
...
+  extends: 'Property',
+
+  properties: [
+    {
+      name: 'adapt',
+      value: function(_, a, p) {
+        var s = typeof a === 'number'   ? String(a)                :
+                a && a.toString         ? a.toString()             :
+                                          ''                       ;
+        return p.trim ? s.trim() : s;
```
Comment (kgrgreer) [inline src/foam/core/types.js]: Why is this p.trim instead of s.trim?

Comment (adamfox-nanopay) [reply src/foam/core/types.js]: p is passed in as a param ... i assumed p.trim is a flag for whether the result should be trimmed, copied from StringPropertyInfo adapt

Comment (adamfox-nanopay) [reply src/foam/core/types.js]: Updated to s.trim()

**Pull request merged**